package/MDAnalysis/topology/tpr/utils.py

```diff
@@ -326,32 +326,63 @@ def do_mtop(data, fver, tpr_resid_from_one=False):
-
-            atom_start_ndx += mt.number_of_atoms()
-            res_start_ndx += mt.number_of_residues()
+        # NOTE: many calculations below appear to be incorrect,
+        # but the test suite says otherwise, or at least is not sensitive...
+        if mt.bonds is not None:
+            curr_bonds = np.tile(mt.bonds, mb.molb_nmol).reshape((mb.molb_nmol * len(mt.bonds), 2))
+            adder = np.repeat(np.arange(0, mt.number_of_atoms() * mb.molb_nmol, mt.number_of_atoms()), 2 * len(mt.bonds)).reshape(-1, 2) + atom_start_ndx
+            curr_bonds += adder
+            bonds.extend(curr_bonds)
+        if mt.angles is not None:
+            curr_angles = np.tile(mt.angles, mb.molb_nmol).reshape((mb.molb_nmol * len(mt.angles), 3))
+            angles.extend(curr_angles)
+        if mt.dihe is not None:
+            curr_dihedrals = np.tile(mt.dihe, mb.molb_nmol).reshape((mb.molb_nmol * len(mt.dihe), 4))
+            dihedrals.extend(curr_dihedrals)
+        if mt.impr is not None:
+            curr_impropers = np.tile(mt.impr, mb.molb_nmol).reshape((mb.molb_nmol * len(mt.impr), 4))
+            impropers.extend(curr_impropers)
+
+        atomids_mol = np.arange(mt.number_of_atoms() * mb.molb_nmol, dtype=np.int32) + atom_start_ndx
+        atomids.extend(atomids_mol)
+        segids.extend(np.repeat(segid, mt.number_of_atoms() * mb.molb_nmol))
+        resids_mol = np.empty(mt.number_of_atoms(), dtype=np.int32)
+        resnames_mol = np.empty(mt.number_of_atoms(), dtype=object)
+        atomnames_mol = np.empty(mt.number_of_atoms(), dtype=object)
+        atomtypes_mol = np.empty(mt.number_of_atoms(), dtype=object)
+        charges_mol = np.empty(mt.number_of_atoms(), dtype=np.float32)
+        masses_mol = np.empty(mt.number_of_atoms(), dtype=np.float32)
+        elements_mol = np.empty(mt.number_of_atoms(), dtype=object)
+        for n in range(mt.number_of_atoms()):
+            resids_mol[n] = mt.atomkinds[n].resid + res_start_ndx
+            resnames_mol[n] = mt.atomkinds[n].resname
+            atomnames_mol[n] = mt.atomkinds[n].name
+            atomtypes_mol[n] = mt.atomkinds[n].type
+            charges_mol[n] = mt.atomkinds[n].charge
+            masses_mol[n] = mt.atomkinds[n].mass
+            elements_mol[n] = mt.atomkinds[n].element_symbol
+        resids_mol = np.tile(resids_mol, mb.molb_nmol)
+        adder = np.repeat(np.arange(0, mt.number_of_residues() * mb.molb_nmol, mt.number_of_residues()), mt.number_of_atoms())
+        resids_mol += adder
+        resids.extend(resids_mol)
+        resnames.extend(np.tile(resnames_mol, mb.molb_nmol))
+        atomnames.extend(np.tile(atomnames_mol, mb.molb_nmol))
+        atomtypes.extend(np.tile(atomtypes_mol, mb.molb_nmol))
+        moltypes.extend([molblock] * (mt.number_of_atoms() * mb.molb_nmol))
+        charges.extend(np.tile(charges_mol, mb.molb_nmol))
+        masses.extend(np.tile(masses_mol, mb.molb_nmol))
+        elements.extend(np.tile(elements_mol, mb.molb_nmol))
```

Comment:
also, as a general comment--this is kind of partial adoption of NumPy, since we could try to embrace NumPy data structures more completely at the higher level rather than extending, but this is where I landed based on the initial bottlenecks from `line_profiler`

Reply:
I think ultimately we need to move TPR parsing into being compiled.  I was holding off on this on the hope that the TNG format (which was meant to hold topology info etc) might be a good replacement, but I don't think that's going to happen.  So I'd probably take these nice improvements and ultimately we might have to cythonize this module.